File: cpp/Structure.cpp
//
// Created by Tamir Eliraz on 11/2/2024.
//

#include "Structure.h"

namespace Sim {
    const Geography::Vector2f &Structure::getLocation() const {
        return location;
    }

    float Structure::getAmount() const {
        return amount;
    }

    Structure &Structure::setAmount(float newAmount) {
        this->amount = newAmount;
        return *this;
    }
} // Sim

File: java/src/main/java/Structure.java
import lombok.Getter;
import org.jetbrains.annotations.NotNull;

@Getter
public abstract class Structure extends SimObject {
    protected float amount;
    private final @NotNull Vector2f location;
    
    protected Structure(@NotNull String name, @NotNull Vector2f location, float amount) {
        super(name);
        this.location = location;
        this.amount = amount;
    }
    
    @Override
    public void update() { }
    
    public void deposit(float supply) { }
    
    public float withdraw(float request) { return 0.0f; }
    
    protected float withdrawHelper(float request) {
        float res;
        if (request > getAmount()) {
            res = getAmount();
            amount = 0;
        } else {
            res = request;
            amount -= request;
        }
        return res;
    }
}
